src/components/common/dialog.tsx

```diff
@@ -0,0 +1,1378 @@
+                this.#dialog._rename_validation_state(this.#path, j + 1, j);
+            this.set(toSpliced(val, index, 1) as T);
+        }
+    }
+
+    add(item: ArrayElement<T>) {
+        const val = this.get();
+        if (Array.isArray(val)) {
+            this.set(val.concat(item) as T);
+        }
+    }
+
+    sub<K extends keyof T>(tag: K, update_func?: ((val: T[K]) => void) | undefined): DialogValue<T[K]> {
+        return new DialogValue<T[K]>(
+            this.#dialog,
+            () => this.get()[tag],
+            (val, is_aux) => {
+                const container = this.get();
+                if (Array.isArray(container) && typeof tag == "number")
+                    this.#setter(toSpliced(container, tag, 1, val) as T, is_aux);
+                else
+                    this.#setter({ ...container, [tag]: val }, is_aux);
+                if (update_func)
+                    update_func(val);
+            },
+            this.#path ? this.#path + "." + String(tag) : String(tag)
+        );
+    }
+
+    at<TT extends T>(witness: TT): DialogValue<TT> {
+        cockpit.assert(Object.is(witness, this.get()));
+        return new DialogValue<TT>(
+            this.#dialog,
+            () => this.get() as TT,
+            (val, is_aux) => {
+                this.#setter(val, is_aux);
+            },
+            this.#path,
+        );
+    }
+
+    validate(func: (val: T) => string | undefined): void {
+        const val = this.get();
+        if (!this.#dialog._probe_validation_cache(this.#path, val)) {
+            const result = func(val);
+            console.log("VALIDATE", this.#path, result);
+            this.#dialog._set_validation_result(this.#path, val, result);
+        }
+    }
+
+    validate_async(debounce: number, func: (val: T) => Promise<string | undefined>): void {
+        const val = this.get();
+        if (!this.#dialog._probe_validation_cache(this.#path, val)) {
+            console.log("START VALIDATE ASYNC", this.#path, val);
+            this.#dialog._set_validation_timeout(
+                this.#path,
+                val,
+                debounce,
+                () => {
+                    console.log("BEGIN VALIDATE ASYNC", this.#path, val);
+                    const prom =
+                        func(val)
+                                .catch(
+                                    ex => {
+                                        console.error(ex);
+                                        return undefined;
+                                    }
+                                )
+                                .then(
+                                    result => {
+                                        if (this.#dialog._validation_state_is_current(this.#path, prom)) {
+                                            console.log("DONE VALIDATE ASYNC", this.#path, result);
+                                            this.#dialog._set_validation_result(this.#path, val, result);
```

Comment:
These 5 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-6408b0c9-20251216-134315-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)